internal/datacoord/snapshot_manager.go

```diff
@@ -0,0 +1,712 @@
+// This is the internal implementation of restoreSnapshotByCopy from services.go.
+// The jobID must be pre-allocated by the caller.
+func (sm *snapshotManager) createRestoreJob(
+	ctx context.Context,
+	targetCollection int64,
+	channelMapping map[string]string,
+	partitionMapping map[int64]int64,
+	snapshotData *SnapshotData,
+	jobID int64,
+) error {
+	log := log.Ctx(ctx).With(
+		zap.String("snapshotName", snapshotData.SnapshotInfo.GetName()),
+		zap.Int64("targetCollectionID", targetCollection),
+		zap.Int64("jobID", jobID),
+		zap.Any("channelMapping", channelMapping),
+		zap.Any("partitionMapping", partitionMapping),
+	)
+
+	// Validate which segments exist in meta
+	validSegments := make([]*datapb.SegmentDescription, 0, len(snapshotData.Segments))
+	for _, segDesc := range snapshotData.Segments {
+		sourceSegmentID := segDesc.GetSegmentId()
+		segInfo := sm.meta.GetSegment(ctx, sourceSegmentID)
+		if segInfo == nil {
+			log.Warn("source segment not found in meta, skipping",
+				zap.Int64("sourceSegmentID", sourceSegmentID))
+			continue
+		}
+		validSegments = append(validSegments, segDesc)
+	}
+
+	// Allocate target segment IDs
+	// AllocN returns (start, end, error), where end = start + count
+	targetSegmentIDStart, _, err := sm.allocator.AllocN(int64(len(validSegments)))
+	if err != nil {
+		log.Error("failed to allocate segment IDs", zap.Error(err))
+		return err
+	}
+
+	// Create ID mappings and pre-register target segments
+	idMappings := make([]*datapb.CopySegmentIDMapping, 0, len(validSegments))
+	totalRows := int64(0)
+	targetSegments := make(map[int64]*SegmentInfo, len(validSegments))
+	for i, segDesc := range validSegments {
+		sourceSegmentID := segDesc.GetSegmentId()
+		targetSegmentID := targetSegmentIDStart + int64(i)
+		totalRows += segDesc.GetNumOfRows()
+
+		// Apply partition mapping
+		sourcePartitionID := segDesc.GetPartitionId()
+		targetPartitionID, ok := partitionMapping[sourcePartitionID]
+		if !ok {
+			// L0 segments may not have partition mapping
+			if segDesc.GetSegmentLevel() == datapb.SegmentLevel_L0 {
+				targetPartitionID = -1
+			} else {
+				err := merr.WrapErrServiceInternal(
+					fmt.Sprintf("partition mapping not found for segment: sourceSegmentID=%d, sourcePartitionID=%d",
+						sourceSegmentID, sourcePartitionID))
+				log.Error("partition mapping missing", zap.Error(err))
+				return err
+			}
+		}
+
+		idMappings = append(idMappings, &datapb.CopySegmentIDMapping{
+			SourceSegmentId: sourceSegmentID,
+			TargetSegmentId: targetSegmentID,
+			PartitionId:     targetPartitionID,
+		})
+
+		// Apply channel mapping
+		targetChannelName, ok := channelMapping[segDesc.GetChannelName()]
+		if !ok {
+			err := merr.WrapErrServiceInternal(
+				fmt.Sprintf("channel mapping missing for channel: %s", segDesc.GetChannelName()))
+			log.Error("channel mapping not found", zap.Error(err))
+			return err
+		}
+
+		// Prepare positions with correct channel names
+		startPos := segDesc.GetStartPosition()
```

Comment:
why we need to inherit position from old segemnt?
those position might not already useless?

Thought this is only useful for growing segment. but now we are more like doing import.

Don't think we even need to select the same chanel